README.md

```diff
@@ -65,147 +65,227 @@ TechInfo lookup is looking at Toyota's Techinfo site (payment required, minimum
+ - Write data by identifier 0x202 00000000000000000000000000000000
+
+Upload payload...
+ - Request download
+ - Transfer data 0
+ - Transfer data 1
+ - Transfer data 2
+ - Transfer data 3
+
+Verify payload...
+ - Routine control 0x10f0 OK!
+
+Trigger payload...
+
+Dumping keys...
+100%|████████████████████████| 448/448 [00:00<00:00, 14293.36it/s]
+
+ECU_MASTER_KEY    82667ef509b9f07a134aaf89d4973c68
+SecOC Key (KEY_4) 0123456789abcdef0123456789abcdef
+
+SecOC key written to param successfully!
+comma@comma-71b93b83:/data/openpilot/secoc$ 
+```
+This 32 digit hexadecimal number is your key.
+> SecOC Key (KEY_4) **0123456789abcdef0123456789abcdef**
+
+Email yourself the key so that you don't need to extract it again.
+
+Proceed to Step 5.
+
+> [!WARNING]
+> It's theoretically possible for someone to hack your car with the key _if the hacker has physical access to your car_. You don't need to protect the key like it's your bank password, but still don't post it on Discord.
+
+### Step 4. Debugging
+
+#### 4-1. `panda.python.uds.MessageTimeoutError: timeout waiting for response`
+Turn off the car, put it back into `Not Ready to Drive` mode, and then try again.
+
+#### 4-2. `panda.python.uds.InvalidServiceIdError: invalid response service id: 0x50` or similar
+Turn off the car, put it back into `Not Ready to Drive` mode, and then try again.
+
+#### 4-3. `Can't read application software identification. Please cycle ignition.`
+Turn off the car, put it back into `Not Ready to Drive` mode, and then try again.
+
+#### 4-4. `Unexpected application version!`
+4-4-1. Open the script for editing.
+```sh
+nano -l /data/openpilot/secoc/extract_keys.py
+```
+
+4-4-2. Comment out lines 78 and 93 by adding a `#` at the beginning of each line.
+```python
+if app_version not in APPLICATION_VERSIONS:
+    print("Unexpected application version!", app_version)
+#    exit(1)
+```
+```python
+if bl_version != APPLICATION_VERSIONS[app_version]:
+    print("Unexpected bootloader version!", bl_version)
+#    exit(1)
+```
+
+4-4-2. Save and exit the editor (`Ctrl+X`, then `Y`, then `Enter`).
+
+4-4-3. Run the script again.
+```sh
+./extract_keys.py
+```
+
+#### 4-5. Still doesn't work?
+Turn off the car, unplug everything, plug them back in, and try Step 3 again.
+
+
+### Step 5. Write the security key to params
+5-1. Write the key.
+```sh
+echo -n "your key here" > /data/params/d/SecOCKey
+```
+For example,
+```sh
+echo -n "0123456789abcdef0123456789abcdef" > /data/params/d/SecOCKey
+```
+5-2. Disable updates.
+```sh
+echo -n "1" > /data/params/d/DisableUpdates
+```
+
+### Step 6. Fingerprinting (if the car is not recognized)
+6-1. Follow the fingerprinting guide to get the ECU codes: https://github.com/commaai/openpilot/wiki/Fingerprinting
+
+6-2. Add the ECU codes to `fingerprints.py`.
+
+6-2-1. Open the file for editing. 
+```sh
+nano /data/openpilot/selfdrive/car/toyota/fingerprints.py
+```
+
+6-2-2. Scroll down to the `CAR.TOYOTA_RAV4_PRIME` section.
```

Comment:
Is this still valid?

Reply:
I might suggest giving it as an example. And/or since it's just two vehicles for now, maybe add a Sienna version or shout out. But if/when more is added, this could be generalized.

Reply:
I'd prefer that someone who's more familiar expand this section

Reply:
Let's just leave this in as-is. I actually would prefer people switch over to https://github.com/hardcoded-fp/openpilot and forget even trying to edit files and stuff. They will eventually be picked up automatically by comma's system and their fingerprints will make it up.